packages/sdk/src/libs/tables/base/base.ts

```diff
@@ -0,0 +1,37 @@
+import assert from "assert";
+import { exists } from "../../utils";
+import type { Store, MakeId, MaybeId, HasId } from "../../index.d";
+
+export default function Table<I, D>(config: { makeId: MakeId<I, D>; type: string }, store: Store<I, D>) {
```

Comment:
I think the design of the store and table are quite nice. It think the types are pretty well articulated and organized. However, I don't quite understand why we need both. One seems to be slightly higher-level than the other, but is there any reason for that? Do we use the store independently?

One (admittedly unlikely issue depending on our ingestion architecture) with keeping them separate is that _sometimes_ you might want the higher-level functions to act in a highly consistent way (from a remote database perspective).

There are two examples of possible issues you might see with a remote database:

```ts
  async function create(data: D & MaybeId<I>) {
    // Doesn't exist.
    const id = exists(data.id) ? data.id : makeId(data);
    // Someone else creates it just after we checked it.
    assert(!(await has(id)), `${type} exists`);
    // We overwrite their newly created datum!!
    return set({ id, ...data });
  }
```

Similarly:
```ts
  async function update(id: I, data: Partial<D>) {
    const got = await get(id);
    // Data changes just after we retrieve the data for id.
    // We revert the other change because `got` is out of date by the time we call set!!
    return set({ ...got, ...data });
  }
```

If table is aware of the underlying database implementation, it can use the tools from that db to create consistent transactions: see the transaction object from google cloud datastore -- https://googleapis.dev/nodejs/datastore/latest/Transaction.html.

Reply:
yea these are great points. ill try to respond to each.
1. data consistency - the implicit assumption with these tables is that there should only ever be one process writing, though many can read. with multiple writers you run into the exact problem you are illustrating. a "process" can also be thought of an ordered stream of mutations (calls to table), as long as each mutation is ordered, the reads will be consistent within each function call.
2. database specific optimizations - you are right that there is a lot to take advantage of with google datastore, or any other db, im trying to avoid lock in by only using the most primitive CRUD style methods.  This is less performant, but im favoring flexibility over performance. Its not possible to avoid all DB specific calls, but the idea is to create a consistent interface in front of tech specific code.
3. store vs table: the store is technology specific but data agnostic, while the table is data specific but technology agnostic. The table acts as your interface to data. Its not perfect, but the idea is within your application you could easily switch or add new database "stores" without changing much if any business logic that references tables.  In my exp this flexibility unlocks a lot of power. If these things were combined you would find yourself repeating a lot more code than necessary.
4. using stores independently - you probably wouldnt use it independently unless you just wanted a very simple KV interface to a database. the interface though is what allows you to easily add new database adapters. You can imagine quickly adding a mongodb, redis, or even local storage adapter ( for frontend), since the interface is so simple, and all of a sudden the rest of your app can consume it with very minimal work outside the store.

Reply:
> 1. data consistency - the implicit assumption with these tables is that there should only ever be one process writing, though many can read. with multiple writers you run into the exact problem you are illustrating. a "process" can also be thought of an ordered stream of mutations (calls to table), as long as each mutation is ordered, the reads will be consistent within each function call.

This seems like a pretty big assumption we're making if this is the storage layer for all database interactions. I could imagine us violating this if, for instance, we manually need to modify the database outside of the ingestion pipeline.

> database specific optimizations

I'm less worried about optimizations, and more worried about correctness.

> using stores independently - you probably wouldnt use it independently unless you just wanted a very simple KV interface to a database. the interface though is what allows you to easily add new database adapters. You can imagine quickly adding a mongodb, redis, or even local storage adapter ( for frontend), since the interface is so simple, and all of a sudden the rest of your app can consume it with very minimal work outside the store.

Maybe a different question is whether we imagine other interfaces to the store other than Table? Put another way, is there a reason for the extra layer of abstraction since the table has fewer methods and could be simpler to implement.

Reply:
> his seems like a pretty big assumption we're making if this is the storage layer for all database interactions. I could imagine us violating this if, for instance, we manually need to modify the database outside of the ingestion pipeline.

this is a drawback of the design. in practice you can usually get away with mutating directly to database without issue, this is dangerous though in any system, particularly if you have a caching layer. the right way to do this is to expose privileged calls through your service to touch the database, this way you can validate data and guard against changes which should not be allowed, update caches and also avoid inconsistent state.  

> Maybe a different question is whether we imagine other interfaces to the store other than Table? Put another way, is there a reason for the extra layer of abstraction since the table has fewer methods and could be simpler to implement.

I think i might be missing some of the nuances to your thinking here.  Maybe what is missing is that the tables currently are just passthroughs of a store. Eventually tables could be more complicated than stores and have specific calls to change data ( you can imagine a balances table which validates withdrawals against user balance). This business logic can be unit tested with a js map store in memory, the assumption is then it will work with any other store without having to deploy a db for the tests.  In many cases someone can update the table logic without knowing anything specific about how to use a specific database. 

Hopefully this makes some sense, though would be happy to discuss offline as well!